### PR TITLE
Caaspv4 deploy using terraform against Crowbar deployment (SOC-11160)

### PR DESCRIPTION
**Pull request by mbelur**
The change https://github.com/SUSE-Cloud/automation/pull/3851
supported deploying caaspv4 using terraform against ardana
based SOC deployments.

It has been extended to support deploying caaspv4 against
crowbar based virtual SOCC deployments in ECP.

### CODE REVIEW

File: scripts/jenkins/cloud/ansible/roles/deploy_caasp_terraform/tasks/deploy_caasp.yml [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+#
+---
+- name: Management Machine | Update resolv.conf
+  shell:
+    echo "nameserver {{ dns_server_ip1 }}" | sudo tee -a /etc/resolv.conf
```
Comment (djoreilly) [inline scripts/jenkins/cloud/ansible/roles/deploy_caasp_terraform/tasks/deploy_caasp.yml:19]: This change gets lost when the lease is renewed. I did a PR for ardana so neutron's dhcp server pushes it out instead.

Comment (mbelur) [reply scripts/jenkins/cloud/ansible/roles/deploy_caasp_terraform/tasks/deploy_caasp.yml]: Updated

**Comment by mbelur** [comment]
crowbar deploy with caaspv4: https://ci.suse.de/job/openstack-crowbar/920
ardana deploy with caaspv4: https://ci.suse.de/job/openstack-ardana/639

File: scripts/jenkins/cloud/ansible/roles/deploy_caasp_terraform/defaults/main.yml [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+dns_server_ip2: "10.84.2.21"
+image_url: "http://ibs-mirror.prv.suse.net/install/SLE-15-SP1-JeOS-QU2/SLES15-SP1-JeOS.x86_64-15.1-OpenStack-Cloud-QU2.qcow2"
+image_path: "{{ soc_home }}/SLES15-SP1-JeOS.x86_64-15.1-OpenStack-Cloud-QU2.qcow2"
+image_name: "SLE-15-SP1-JeOS-QU2"
+timestamp: ''
+internal_net: "{{'caasp-int-net' + '-' + timestamp }}"
+internal_subnet: "{{'caasp-int-subnet' + '-' + timestamp}}"
+keyname: "caaspmanagementkey"
+lb_tf_path: "{{ caasp_openstack_path }}/load-balancer.tf"
+master_size: "m1.medium"
```
Comment (djoreilly) [inline scripts/jenkins/cloud/ansible/roles/deploy_caasp_terraform/defaults/main.yml:33]: The ECP large flavor gives a 80GB disk and inside the VM it's partitioned so the root fs is only 51GB. The m1.medium requires 40GB, so can only fit one master on a ECP VM as disk over subscription is disabled. But the master is only using ~5GB.  Maybe create new flavor with the same parameters as m1.medium but with only 10GB for disk. Same goes for workers.

File: scripts/jenkins/cloud/ansible/roles/deploy_caasp_terraform/tasks/deploy_caasp.yml [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+    line: "{{ item.line }}"
+  loop:
+    - { dest: "{{ tfvars_path }}", regexp: '^image_name = "', line: 'image_name = "{{ image_name }}"' }
+    - { dest: "{{ tfvars_path }}", regexp: '^internal_net = "', line: 'internal_net = "{{ internal_net }}"' }
+    - { dest: "{{ tfvars_path }}", regexp: '^internal_subnet = "', line: 'internal_subnet = "{{ internal_subnet }}"' }
+    - { dest: "{{ tfvars_path }}", regexp: '^external_net = "', line: 'external_net = "{{ external_net }}"' }
+    - { dest: "{{ tfvars_path }}", regexp: '^stack_name = "', line: 'stack_name = "{{stack_name}}"' }
+    - { dest: "{{ tfvars_path }}", regexp: '^subnet_cidr = "', line: 'subnet_cidr = "{{ subnet_cidr }}"' }
+    - { dest: "{{ tfvars_path }}", regexp: '^master_size = "', line: 'master_size = "{{ master_size }}"' }
+    - { dest: "{{ tfvars_path }}", regexp: '^worker_size = "', line: 'worker_size = "{{ worker_size }}"' }
```
[the remark this answers is not in the document]
Comment (mbelur) [reply scripts/jenkins/cloud/ansible/roles/deploy_caasp_terraform/tasks/deploy_caasp.yml]: updated

File: scripts/jenkins/cloud/ansible/roles/deploy_caasp_terraform/defaults/main.yml (hunk 2 above)
Comment (mbelur) [reply scripts/jenkins/cloud/ansible/roles/deploy_caasp_terraform/defaults/main.yml]: Updated to use a custom flavor.

File: scripts/jenkins/cloud/ansible/roles/deploy_caasp_terraform/tasks/deploy_caasp.yml [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+  set_fact:
+    count: "{{ count | default(0) | int + 1 }}"
+  when: "'node successfully joined the cluster' in item"
+  loop: "{{ skuba_join_commands_results.results|map(attribute='stdout')|list }}"
+
+- name: Management Machine | Fail when the expected number of nodes do not join the cluster
+  fail:
+    msg: "Not all expected nodes have successfully joined the cluster"
+  when: "count | int != total_join | int"
+
```
Comment (djoreilly) [inline scripts/jenkins/cloud/ansible/roles/deploy_caasp_terraform/tasks/deploy_caasp.yml:312]: I did a deployment with num_masters=2 and num_workers=2.
https://ci.suse.de/blue/organizations/jenkins/openstack-crowbar/detail/openstack-crowbar/923/pipeline/158
It failed at this point. For some reason skuba printed `node successfully joined the cluster` twice. But `skuba cluster status` looks alright.

```
sles@caasp-management-20200417095828:~/caasp/deployment/openstack/auto-caasp-20200417095828-cluster> cat skuba_join.log 
[join] applying states to new node
[join] applying states to new node
[join] node successfully joined the cluster
[join] applying states to new node
[join] node successfully joined the cluster
sles@caasp-management-20200417095828:~/caasp/deployment/openstack/auto-caasp-20200417095828-cluster> skuba cluster status
NAME                                       STATUS   ROLE     OS-IMAGE                              KERNEL-VERSION           KUBELET-VERSION   CONTAINER-RUNTIME   HAS-UPDATES   HAS-DISRUPTIVE-UPDATES   CAASP-RELEASE-VERSION
caasp-master-auto-caasp-20200417095828-0   Ready    master   SUSE Linux Enterprise Server 15 SP1   4.12.14-197.37-default   v1.16.2           cri-o://1.16.1      <none>        <none>                   <none>
caasp-master-auto-caasp-20200417095828-1   Ready    master   SUSE Linux Enterprise Server 15 SP1   4.12.14-197.37-default   v1.16.2           cri-o://1.16.1      no            no                       4.1.2
caasp-worker-auto-caasp-20200417095828-0   Ready    <none>   SUSE Linux Enterprise Server 15 SP1   4.12.14-197.37-default   v1.16.2           cri-o://1.16.1      no            no                       4.1.2
caasp-worker-auto-caasp-20200417095828-1   Ready    <none>   SUSE Linux Enterprise Server 15 SP1   4.12.14-197.37-default   v1.16.2           cri-o://1.16.1      no            no                       4.1.2

```

Comment (djoreilly) [reply scripts/jenkins/cloud/ansible/roles/deploy_caasp_terraform/tasks/deploy_caasp.yml:312]: It's also strange that skuba did not set the role for the workers. But `--role worker` was in the command line. But that is a skuba problem.

Comment (mbelur) [reply scripts/jenkins/cloud/ansible/roles/deploy_caasp_terraform/tasks/deploy_caasp.yml:312]: I'm not able to reproduce the issue:

I have tried using num_masters=3 in this example run: 
  https://ci.suse.de/job/openstack-caaspv4/5/console
I have tried using num_masters=2 in this example run: 
  https://ci.suse.de/job/openstack-caaspv4/6/console

Both the test runs passed.

Not sure why it has failed in your case. I'm building a crowbar environment in ECP to see if I can reproduce this issue there.

Yeah about the worker role.. I did notice that and it's a skuba issue.

Comment (djoreilly) [reply scripts/jenkins/cloud/ansible/roles/deploy_caasp_terraform/tasks/deploy_caasp.yml:312]: Maybe this task is not totally reliable. The next task counts the "Ready" nodes from `skuba cluster status` - maybe that's enough?

Comment (mbelur) [reply scripts/jenkins/cloud/ansible/roles/deploy_caasp_terraform/tasks/deploy_caasp.yml:312]: Updated this task to just print the message as we check the status in the next task.

**Comment by djoreilly** [comment]
LGTM. Thanks.

**Pull request merged**